### PR TITLE
09-implemented-shopping-cart-Cart-item

### PR DESCRIPTION
**Pull request by RostyslavOnysh**

### CODE REVIEW

File: src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public CartItemResponseDto update(UpdateQuantityDto updateQuantityDto, Long id) {
+        CartItem cartItem = cartItemRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find Entity by ID" + id));
+        cartItem.setQuantity(updateQuantityDto.getQuantity());
+        return mapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public CartItemResponseDto getAllCartItemsForShoppingCart(Long shoppingCartId,
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java:56]: I don't get what this method is for. For cart item service save,getById, and delete should be enough - these methods would be used in shopping cart service

File: src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    private final BookRepository bookRepository;
+    private final UserService userService;
+
+    @Override
+    public CartItemResponseDto save(CartItemRequestDto cartItemRequestDto) {
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(bookRepository.getById(cartItemRequestDto.getBookId()));
+        cartItem.setQuantity(cartItemRequestDto.getQuantity());
+        Long id = userService.getAuthenticated().getId();
+        cartItem.setShoppingCart(shoppingCartRepository.getById(id));
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java:36]: use mapper to transform request dto to CartItem model

File: src/main/java/spring/boot/bookstore/service/shoppingcart/ShoppingCartService.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package spring.boot.bookstore.service.shoppingcart;
+
+import spring.boot.bookstore.dto.cartitem.CartItemRequestDto;
+import spring.boot.bookstore.dto.cartitem.CartItemResponseDto;
+import spring.boot.bookstore.dto.shoppingcart.ShoppingResponseDto;
+import spring.boot.bookstore.model.ShoppingCart;
+
+public interface ShoppingCartService {
+
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/shoppingcart/ShoppingCartService.java:9]:
```suggestion
```

File: src/main/java/spring/boot/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                responseDto.setCartItems(shoppingCart.get().getCartItems().stream()
+                        .map(cartItemMapper::toDto)
+                        .collect(Collectors.toSet()));
+                return responseDto;
+            } else {
+                return null;
+            }
+        } else {
+            return null;
+        }
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:64]: bad practice

File: src/main/java/spring/boot/bookstore/repository/shoppingcart/ShoppingCartRepository.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package spring.boot.bookstore.repository.shoppingcart;
+
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import spring.boot.bookstore.model.ShoppingCart;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    Optional<ShoppingCart> getUserById(Long id);
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/repository/shoppingcart/ShoppingCartRepository.java:10]:
```suggestion
    Optional<ShoppingCart> getById(Long id);
```

File: src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java (hunk 2 above)
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java:36]: foreign key constraint error may be occuring here - what if getById returns null?

**Review by okuzan — CHANGES_REQUESTED** [review]
It seems like there is a lot of work that needs to be done in the services. If your application is not functioning properly, it is recommended to post a thread in the chat section discussing the issue and then provide a working solution for review. Service methods should not be overly complex. Try delegating some of the work to mappers and utility methods in entities that are responsible for updating relationships.

File: src/main/java/spring/boot/bookstore/repository/cartitem/CartItemRepository.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+import spring.boot.bookstore.model.CartItem;
+
+@Repository
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @Query("SELECT c FROM CartItem c WHERE c.shoppingCart.id =:shoppingCartId")
+    Set<CartItem> findCartItemByShoppingCartId(@Param("shoppingCartId") Long shoppingCart);
```
Comment (Ivan95kos) [inline src/main/java/spring/boot/bookstore/repository/cartitem/CartItemRepository.java:13]: such simple queries should be handled by proper method naming, please recheck work without them and fix just naming method if required

File: src/main/java/spring/boot/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import spring.boot.bookstore.repository.cartitem.CartItemRepository;
+import spring.boot.bookstore.repository.shoppingcart.ShoppingCartRepository;
+import spring.boot.bookstore.service.cartitem.CartItemService;
+import spring.boot.bookstore.service.shoppingcart.ShoppingCartService;
+import spring.boot.bookstore.service.user.UserService;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+
```
[the remark this answers is not in the document]
Comment (Ivan95kos) [reply src/main/java/spring/boot/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:19]: not fixed

File: src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import spring.boot.bookstore.repository.BookRepository;
+import spring.boot.bookstore.repository.cartitem.CartItemRepository;
+import spring.boot.bookstore.repository.shoppingcart.ShoppingCartRepository;
+import spring.boot.bookstore.service.cartitem.CartItemService;
+import spring.boot.bookstore.service.user.UserService;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
+
```
[the remark this answers is not in the document]
Comment (Ivan95kos) [reply src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java:27]: not fixed

File: src/main/java/spring/boot/bookstore/repository/BookRepository.java [hunk 9]
```diff
@@ -13,9 +14,10 @@ public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificat
     List<Book> findAllByTitleContainsIgnoreCase(String title);
 
     @Query("FROM Book b JOIN b.categories c WHERE c.id =:categoryId")
-    List<Book> findByCategoryId(Long categoryId, Pageable pageable);
+    List<Book> findByCategoryId(@Param("categoryId") Long categoryId,
```
Comment (Ivan95kos) [inline src/main/java/spring/boot/bookstore/repository/BookRepository.java:17]: such simple queries should be handled by proper method naming, please recheck work without them and fix just the naming method if required

**Review by Ivan95kos — CHANGES_REQUESTED** [review]
Please, fix the previous comments

File: src/main/java/spring/boot/bookstore/service/cartitem/CartItemService.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.Set;
+import spring.boot.bookstore.dto.cartitem.CartItemRequestDto;
+import spring.boot.bookstore.dto.cartitem.CartItemResponseDto;
+import spring.boot.bookstore.dto.cartitem.UpdateQuantityDto;
+import spring.boot.bookstore.model.CartItem;
+import spring.boot.bookstore.model.User;
+
+public interface CartItemService {
+
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/cartitem/CartItemService.java:11]:
```suggestion
```

File: src/main/java/spring/boot/bookstore/service/user/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -44,5 +46,13 @@ public UserResponseDto register(UserRegistrationRequestDto request)
         user.setRoles(defaultUserRoleSet);
         return userMapper.toDto(userRepository.save(user));
     }
+
+    @Override
+    public User getAuthenticated() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        return userRepository.findByEmail(authentication.getName()).orElseThrow(
+                () -> new RuntimeException("Can`t find user with according email"
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/user/impl/UserServiceImpl.java:54]: don't use generic runtime exception

File: src/main/java/spring/boot/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                .orElseGet(() -> registerNewCart(authenticatedUser));
+        Long id = shoppingCart.getId();
+        ShoppingCartResponseDto shoppingCartDto = new ShoppingCartResponseDto();
+        shoppingCartDto.setId(id);
+        shoppingCartDto.setUserId(authenticatedUser.getId());
+        shoppingCartDto.setCartItems(cartItemService.findByShoppingCartId(id));
+        return shoppingCartDto;
+    }
+
+    private ShoppingCart registerNewCart(User user) {
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:43]: duplicated method in two services

File: src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        CartItem cartItem = cartItemRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find cart by ID : " + id));
+        cartItem.setQuantity(updateQuantityDto.getQuantity());
+        return mapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public void delete(Long id) {
+        cartItemRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can`t find cart by ID : " + id));
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java:62]:
```suggestion
```

File: src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public Set<CartItemResponseDto> findByShoppingCartId(Long id) {
+        return cartItemRepository.findCartItemByShoppingCartId(id).stream()
+                .map(mapper::toDto)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public CartItemResponseDto update(UpdateQuantityDto updateQuantityDto, Long id) {
```
Comment (okuzan) [inline src/main/java/spring/boot/bookstore/service/cartitem/impl/CartItemServiceImpl.java:52]: check if quantity is positive

**Review by okuzan — CHANGES_REQUESTED** [review]

**Pull request merged**